### PR TITLE
fix: update retry strategy for mutation calls to handle aborted transactions

### LINKED ISSUE #1133: database.batch does not retry aborted transactions
The standard example for writing data with mutations uses `database.batch`: https://cloud.google.com/spanner/docs/getting-started/python#write-data-with-mutations

`database.batch` however does not automatically retry the transaction if it is aborted by Spanner. This causes errors if you try to use this method to insert a large amount of data, or if there are lock contentions on the data that you insert.

Either:
1. The sample(s) should be updated to show how to use `run_in_transaction` to use mutations.
2. And/or: `database.batch` should also automatically retry aborted transactions.

### PR DESCRIPTION
**Pull request by aakashanandg**
## Updating retry strategy for mutation calls to handle aborted transactions
This PR updates the retry strategy for mutation calls to handle aborted transactions more effectively. Previously, the retry mechanism didn't handle certain edge cases for aborted transactions, leading to failures. The updated strategy ensures retries in these scenarios to improve the robustness of the mutation operations.
## Test Results:
All unit tests related to the retry logic and mutation operations pass successfully.

Fixes #1133

### CODE REVIEW

File: google/cloud/spanner_v1/batch.py [hunk 1]
```diff
@@ -293,7 +305,9 @@ def group(self):
         self._mutation_groups.append(mutation_group)
         return MutationGroup(self._session, mutation_group.mutations)
 
-    def batch_write(self, request_options=None, exclude_txn_from_change_streams=False):
+    def batch_write(
```
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/batch.py:308]: Understood. In that case, we can bypass the retry behavior for this operation.

File: google/cloud/spanner_v1/_helpers.py [hunk 2]
```diff
@@ -466,13 +470,19 @@ def _retry(
...
     beforeNextRetry=None,
+    deadline=None,
 ):
     """
-    Retry a function with a specified number of retries, delay between retries, and list of allowed exceptions.
+    Retry a specified function with different logic based on the type of exception raised.
+
+    If the exception is of type google.api_core.exceptions.Aborted,
+    apply an alternate retry strategy that relies on the provided deadline value instead of a fixed number of retries.
+    For all other exceptions, retry the function up to a specified number of times.
```
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/_helpers.py:480]: If the exception is of type `Aborted`, it will trigger the custom retry strategy. However, this will only happen if the user has specified this exception in the `allowed_exceptions` map. If not, the exception will not be retried. For the batch API use case, we will explicitly allow this exception to be retried.

File: google/cloud/spanner_v1/batch.py [hunk 3]
```diff
@@ -372,3 +393,8 @@ def _make_write_pb(table, columns, values):
     return Mutation.Write(
         table=table, columns=columns, values=_make_list_value_pbs(values)
     )
+
+
+def no_op_handler(exc):
+    # No-op (does nothing)
+    pass
```
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/batch.py:400]: Yes, we can use a no-op lambda for this.

File: tests/unit/test_batch.py [hunk 4]
```diff
@@ -618,6 +651,36 @@ def __init__(self, database=None, name=TestBatch.SESSION_NAME):
     def session_id(self):
         return self.name
 
+    def run_in_transaction(self, fnc):
```
[the remark this answers is not in the document]
Comment (aakashanandg) [reply tests/unit/test_batch.py:654]: Yes, it's not needed here. This change is carried forward from my previous PR where we call the `run_in_transaction` method instead.

File: google/cloud/spanner_v1/_helpers.py [hunk 5]
```diff
@@ -529,6 +547,61 @@ def _metadata_with_leader_aware_routing(value, **kw):
...
+    :param attempts: number of call retries
+    """
+
+    cause = exc.errors[0]
+    now = time.time()
+    if now >= deadline:
+        raise
+
+    delay = _get_retry_delay(cause, attempts)
+    print(now, delay, deadline)
```
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/_helpers.py:571]: I extracted these methods to make them more generic, allowing other clients to reuse the logic instead of it being tightly coupled with the session object.

File: google/cloud/spanner_v1/_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply google/cloud/spanner_v1/_helpers.py:496]: I meant that the `_retry` function and the `_retry_on_aborted_exception` should be completely separated. I don't really see any advantage of combining them, as the actual code that can be shared is minimal, and the API surface of this function is not logical.

E.g. if you have defined `Aborted` as a retryable exception, but you forget to supply a deadline, then all of a sudden it is not retriable. Also, deadline is only used if you add Aborted as a possible retryable error, and is otherwise ignored if you only supply other error codes. Same with retry_count; it is only used for non-Aborted errors. The fact that there are many combinations of input arguments that don't make any sense, is an indication that the function itself should be split.

File: google/cloud/spanner_v1/_helpers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply google/cloud/spanner_v1/_helpers.py:585]: I meant: Remove the `print(...)` line. We should not print debug info in non-test code (and normally also not in test code).

File: google/cloud/spanner_v1/_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/_helpers.py]: Thanks for the clarification. I've implemented the new retry logic as suggested, separating the `_retry` and `_retry_on_aborted_exception` functions. This ensures clearer logic, as combining them led to confusing combinations of parameters that didn't make sense. Now, the retry logic for non-Aborted and Aborted exceptions is more distinct and easier to manage.

File: google/cloud/spanner_v1/batch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/batch.py]: Removed the redundant code as this is no longer required with the new implementation.

File: tests/unit/test_batch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aakashanandg) [reply tests/unit/test_batch.py]: Removed in subsequent commits.

File: google/cloud/spanner_v1/_helpers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aakashanandg) [reply google/cloud/spanner_v1/_helpers.py]: Noted. Removed from subsequent commits.

File: google/cloud/spanner_v1/_helpers.py [hunk 6]
```diff
@@ -460,6 +464,34 @@ def _metadata_with_prefix(prefix, **kw):
     return [("google-cloud-resource-prefix", prefix)]
 
 
+def _retry_on_aborted_exception(
+    func,
+    deadline,
+    allowed_exceptions=None,
```
Comment (olavloite) [inline google/cloud/spanner_v1/_helpers.py:470]: I think that we can simplify this further and just remove `allowed_exceptions` from this function. It should only retry aborted exceptions.

File: google/cloud/spanner_v1/_helpers.py [hunk 7]
```diff
@@ -460,6 +464,34 @@ def _metadata_with_prefix(prefix, **kw):
...
+            continue
+        except Exception as exc:
+            try:
+                retry_result = _retry(func=func, allowed_exceptions=allowed_exceptions)
+                if retry_result is not None:
+                    return retry_result
+                else:
+                    raise exc
+            except Aborted:
+                continue
```
Comment (olavloite) [inline google/cloud/spanner_v1/_helpers.py:492]: I think that we should remove this part entirely. I know that the previous implementation of `Batch` retried this specific `RST_STREAM` error, but that was just a copy-paste from other methods. That error is not relevant for this type of operation.

File: google/cloud/spanner_v1/_helpers.py [hunk 8]
```diff
@@ -473,6 +505,7 @@ def _retry(
     Args:
         func: The function to be retried.
         retry_count: The maximum number of times to retry the function.
+        deadline: This will be used in case of Aborted transactions.
```
Comment (olavloite) [inline google/cloud/spanner_v1/_helpers.py:508]: nit: remove, this is not relevant anymore

File: google/cloud/spanner_v1/batch.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply google/cloud/spanner_v1/batch.py:307]: Please also remove the `**kwargs` addition again from this PR. It would just be confusing if that is added in this PR, when it is not relevant to the actual change in this PR.

**Review by olavloite — APPROVED** [review]
